src/main/java/co/cask/hydrator/plugin/source/batch/SalesforceBatchSource.java

```diff
@@ -0,0 +1,247 @@
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Macro;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.batch.Input;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.batch.BatchRuntimeContext;
+import co.cask.cdap.etl.api.batch.BatchSource;
+import co.cask.cdap.etl.api.batch.BatchSourceContext;
+import co.cask.hydrator.common.LineageRecorder;
+import co.cask.hydrator.plugin.BaseSalesforceConfig;
+import co.cask.hydrator.salesforce.SalesforceSchemaUtil;
+import co.cask.hydrator.salesforce.authenticator.AuthenticatorCredentials;
+import co.cask.hydrator.salesforce.parser.SalesforceQueryParser;
+import com.sforce.ws.ConnectionException;
+import org.apache.hadoop.conf.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.LocalDate;
+import java.util.stream.Collectors;
+import javax.ws.rs.Path;
+
+
+/**
+ * Plugin returns records from Salesforce using provided by user SOQL query.
+ * Salesforce bulk API is used to run SOQL query. Bulk api returns data in batches.
+ * Every batch is processed as a separate split by mapreduce.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(SalesforceBatchSource.NAME)
+@Description("Plugin to read data from Salesforce in batches.")
+public class SalesforceBatchSource extends BatchSource<String, String, StructuredRecord> {
+  static final String NAME = "SalesforceBatchSource";
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceBatchSource.class);
+
+  private final Config config;
+  private Schema schema;
+
+  SalesforceBatchSource(Config config) throws ConnectionException {
+    this.config = config;
+  }
+
+  static final class Config extends BaseSalesforceConfig {
+    @Description("The SOQL query to retrieve results for")
+    @Macro
+    private final String query;
+
+    Config(Configuration conf) {
+        super(null, conf.get(SalesforceConstants.CLIENT_ID), conf.get(SalesforceConstants.CLIENT_SECRET),
+             conf.get(SalesforceConstants.USERNAME), conf.get(SalesforceConstants.PASSWORD),
+             conf.get(SalesforceConstants.LOGIN_URL));
+
+        this.query = conf.get(SalesforceConstants.QUERY);
+    }
+
+    public String getQuery() {
+      return query;
+    }
+
+    @Override
+    public void validate() {
+      super.validate();
+      SalesforceQueryParser.validateQuery(query);
+    }
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws ConnectionException {
+    this.schema = SalesforceSchemaUtil.getSchemaFromQuery(config.getAuthenticatorCredentials(), config.getQuery());
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
```

Comment:
During configure pipeline macros are not yet substituted, you need to do proper checks.
Please see Datastore as example (https://github.com/data-integrations/google-cloud/blob/develop/src/main/java/co/cask/gcp/datastore/source/DatastoreSource.java#L92).
Also create pipeline where all config values that can be macro are macros and check run.

Please note that during pipeline run methods are called in the following order:
1. configurePipeline
2. prepareRun
3. initialize

Reply:
thanks for the detailed explanation and examples!